keps/sig-api-machinery/4050-generic-controlplane/README.md

```diff
@@ -0,0 +1,768 @@
+
+### Goals
+
+- As always: every PR transforms a working system into a working system, and 
+  PRs are of manageable size.
+- Improve factoring of *kube-apiserver* through layering on-top of 
+  `k/generic-controlplane`, reducing complexity through more explicit structure
+  and reduction of code in `k/k`.
+- `k/generic-controlplane` will provide
+  - a `sample-generic-controlplane` binary
+  - a modular, further customizable (in code) library suitable to build a 
+    working kube-based control plane without vendoring k/k.
+- `k/generic-controlplane` will (optionally) include the ability to define 
+  resources by CustomResourceDefinition objects.
+- `k/generic-controlplane` will be able to (optionally) delegate handling of 
+  some kinds of objects to another server, as directed by APIService objects.
+- `k/generic-controlplane` will allow customization (in code) of which generic
+  (native) resources like secrets, configmaps, admission webhooks, RBAC, etc. 
+  are served.
+- `k/generic-controlplane` will not include the definitions of the resources in 
+  Kubernetes for the management of containerized workloads.  For example, the 
+  excluded resources include: nodes, pods, daemonsets, ingresses, services, 
+  persistentvolumes.
+- `k/generic-controlplane` will allow, but not require all-in-one binaries, both 
+  in a *hyperkube*-like subcommand way, and in a all-in-one *k3s* like way.
+
+### Non-Goals
+
+- provide and ship a de-facto standard, full-featured generic-controlplane binary:
+  
+  i.e. this is clearly a library approach and consumer projects will define a 
+  feature set of a control plane. There is no new deliverable beyond a staging 
+  repository with a library and a sample binary only, with clear limited scope 
+  of demonstrating plumbing.
+- change anything noticeable to the user for existing binaries.
+- change compatibility guarantees of (server-side) staging repositories
+- create `k/kube-apiserver` or anything similar, although this work can lead the
+  path by defining package structures suitable for `k/kube-apiserver`.
+
+## Proposal
+
+The desired outcome is a useful new library, while of course keeping everything 
+working during iterative development. Success will be measured by community 
+members saying that the new library is useful to them.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+Project *kube-hyper-mini* wants to maintain a main program that bundles
+1. the subset of *kube-apiserver* that is not concerned with the management of 
+   containerized workloads, 
+2. a single-member etcd cluster, and 
+3. the subset of *kube-controller-manager* that is not concerned with the 
```

Comment:
I like the theme of this KEP! It makes a lot of sense.

One thing I noticed is that `generic-controlplane` takes on a lot of responsibility.  Have we considered breaking it up into separate `generic-apiserver` and `generic-controller-manager` parts?

Reply:
Yes, we can do that. I tried to avoid the name "generic-apiserver" because "genericapiserver" is a thing in the code base.

Reply:
alternative, not sure how pretty that is: generic-controlplane-apiserver+controllers.

Reply:
Good point.  I can't think of anything better `generic-controlplane-apiserver` and similar, at the moment, either.

Reply:
Is `core-apiserver` taken?

Reply:
Another idea: universal-apiserver. Or just „controlplane-apiserver“.

Reply:
I tend to like `generic-controlplane-apiserver/controllers`. It‘s clear what it is.

Reply:
Can we put a comment in the KEP to make it clear that we might not have just one staging repo? I think I'm OK with figuring out the exact split as we go, but I don't want sign off on just a single staging repo in the KEP given that we might want to split it up.

Reply:
done.